File: src/classContent/wk5/sameTree/InvertBinaryTree.java
package classContent.wk5.sameTree;

public class InvertBinaryTree {
    public static void main(String[] args) {
        TreeNode tree1 = new TreeNode(4, new TreeNode(2, new TreeNode(1), new TreeNode(3)), new TreeNode(7, new TreeNode(6), new TreeNode(9)));
        TreeNode tree2 = new TreeNode(2, new TreeNode(1), new TreeNode(3));
        TreeNode tree3 = new TreeNode();
        tree1.printTree();
        invertTree(tree1);
        tree1.printTree();
        
        tree2.printTree();
        invertTree(tree2);
        tree2.printTree();
        
        tree3.printTree();
        invertTree(tree3);
        tree3.printTree();
    }

    public static TreeNode invertTree(TreeNode root) {
        if (root == null)
            return null;

        if (root.left == null && root.right == null)
            return root;
        else {
            TreeNode nodeTmp = new TreeNode();
            if (root.left != null) {
                nodeTmp = invertTree(root.left);
            } else {
                nodeTmp = null;
            }
            if (root.right != null) {
                root.left = invertTree(root.right);
            } else {
                root.left = null;
            }
            root.right = nodeTmp;
        }
        
        return root;
    }
}

File: src/leetcode/2024/07/InvertBinaryTree.py
# https://leetcode.com/problems/invert-binary-tree/
from typing import Optional
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if not root:
            return None
        
        # invert left and right subtrees
        root.left, root.right = self.invertTree(root.right), self.invertTree(root.left)
        return root
    
    def printTree(self, root: Optional[TreeNode]) -> None:
        if not root:
            return
        print(root.val)
        self.printTree(root.left)
        self.printTree(root.right)

sol = Solution()
root = TreeNode(4)
root.left = TreeNode(2)
root.right = TreeNode(7)
root.left.left = TreeNode(1)
root.left.right = TreeNode(3)
root.right.left = TreeNode(6)
root.right.right = TreeNode(9)
# sol.printTree(root)
sol.invertTree(root)
sol.printTree(root)

root = TreeNode(2)
root.left = TreeNode(1)
root.right = TreeNode(3)
# sol.printTree(root)
sol.invertTree(root)
sol.printTree(root)